🔍 Investigate and resolve evaluation failure due to chat template conflict during model loading

Tasks for @claude 

### Description

Running `make evaluate` fails with a `ValueError` caused by an attempt to reapply a chat template to the tokenizer, even though it already includes one. This issue prevents the evaluation pipeline from loading the fine-tuned model and executing SQL prediction evaluation.

---

### Issue Summary

Despite setup_chat_format: false and force_chat_setup: false being explicitly set, the evaluation script still triggers setup_chat_format(model, tokenizer) in model_setup.py. The log confirms that the tokenizer already has a chat_template, leading to the raised ValueError.

The problem likely stems from the logic in ModelSetup.load_trained_model() not properly respecting these configuration flags, causing a redundant call to setup_chat_format() when the base model already includes chat metadata.

---

### Expected Behavior

- When setup_chat_format: false is specified, the evaluation pipeline should not attempt to modify or reapply the chat template.
- The model and adapter should load cleanly and proceed to evaluation without raising a `ValueError`.

---

1. Inspect `ModelSetup.load_trained_model()` and verify whether the logic for `setup_chat_format` and `force_chat_setup` is being properly respected.
2. Ensure conditional handling so that `setup_chat_format()` is only invoked when:
    - `setup_chat_format` is `true`, and 
    - the tokenizer does not already include a chat template.
3.  Handle the deprecation by switching from `setup_chat_format()` to `clone_chat_template()` as suggested by the TRL warning.
4. Add a clear log message confirming when chat setup is skipped due to pre-existing templates.
5. Re-run `make evaluate` to confirm the fix.

---

### Notes

The following warnings from Pydantic appear at the start of the evaluation but are likely unrelated to the core issue:
```
UnsupportedFieldAttributeWarning: The 'repr' attribute with value False...
UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True...
```

These can be deprioritized unless they interfere with model or dataset loading.

### Logs
```
make evaluate
📊 Running evaluation...
Configuration:
 hf:
  model_id: meta-llama/Llama-3.1-8B-Instruct
  token: ${oc.env:HF_TOKEN,null}
  username: chrisjcc
  setup_chat_format: false
  force_chat_setup: false
  upload:
    push_to_hub: true
    upload_adapter: true
    upload_merged: false
    upload_dataset: true
    adapter_repo_suffix: -text-to-sql-adapter
    merged_repo_suffix: -text-to-sql-merged
    dataset_repo_name: text-to-sql-spider-dataset
    author_name: Christian Contreras Campana
    license: apache-2.0
    language: en
    include_training_args: true
    commit_message: Upload supervised-fine-tuned text-to-SQL model
dataset:
  name: b-mc2/sql-create-context
  train_samples: 20000
  test_samples: 2500
  train_dataset_path: data/train_dataset.json
  test_dataset_path: data/test_dataset.json
wandb:
  enable: true
  project: text-to-sql-finetuning
  api_key: ${oc.env:WANDB_API_KEY,null}
training:
  resume_from_checkpoint: false
  output_dir: Llama-3.1-8B-text-to-sql-adapter
  num_train_epochs: 5
  per_device_train_batch_size: 1
  gradient_accumulation_steps: 8
  learning_rate: 5.0e-05
  max_seq_length: 2048
  max_grad_norm: 1.0
  warmup_ratio: 0.03
  logging_steps: 10
  lora_alpha: 32
  lora_dropout: 0.1
  lora_r: 16
evaluation:
  model_path: meta-llama/Llama-3.1-8B-Instruct
  adapter_path: chrisjcc/Llama-3.1-8B-Instruct-text-to-sql-adapter
  num_eval_samples: 1000
  batch_size: 8
  temperature: 0.0
  max_new_tokens: 128
  skip_baseline: false
  num_examples: 3
  save_predictions: true
inference:
  mode: interactive
  max_new_tokens: 512
  temperature: 0.0
  question: null
  context: null
  batch_file: null
  output_file: inference_outputs.jsonl
logging:
  level: INFO
  log_dir: logs

🔹 Skipping local existence check for remote Model directory: meta-llama/Llama-3.1-8B-Instruct
[2025-10-16 15:13:46][src.utils][INFO] - GPU available: 1 device(s)
[2025-10-16 15:13:46][src.utils][INFO] -   Device 0: NVIDIA H200 (150.12 GB)

Loading test dataset...
[2025-10-16 15:13:46][src.data_preparation][INFO] - Loading prepared dataset from /home/chrisjcc/text-to-sql-finetuning/data/test_dataset.json
[2025-10-16 15:13:46][src.data_preparation][INFO] - Successfully loaded 2500 samples
✅ Loaded 2500 samples.

================================================================================
Loading fine-tuned model (adapter: chrisjcc/Llama-3.1-8B-Instruct-text-to-sql-adapter)...
================================================================================
[2025-10-16 15:13:46][src.model_setup][INFO] - Loading trained model with adapter from chrisjcc/Llama-3.1-8B-Instruct-text-to-sql-adapter
[2025-10-16 15:13:46][src.model_setup][INFO] - Detected HuggingFace Hub ID: chrisjcc/Llama-3.1-8B-Instruct-text-to-sql-adapter
[2025-10-16 15:13:47][src.model_setup][INFO] - ✓ Model found on HuggingFace Hub
adapter_config.json: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 943/943 [00:00<00:00, 11.3MB/s]
[2025-10-16 15:13:47][src.model_setup][INFO] - Loading base model: meta-llama/Llama-3.1-8B-Instruct
`torch_dtype` is deprecated! Use `dtype` instead!
[2025-10-16 15:13:47][accelerate.utils.modeling][INFO] - We will use 90% of the memory on device 0 for storing the model, and 10% for the buffer to avoid OOM. You can set `max_memory` in to a higher value to use more memory (at your own risk).
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:02<00:00,  1.56it/s]
[2025-10-16 15:13:50][src.model_setup][INFO] - ✓ Base model loaded (vocab size: 128256)
[2025-10-16 15:13:50][src.model_setup][INFO] - Loading tokenizer from checkpoint...
tokenizer_config.json: 50.5kB [00:00, 16.9MB/s]
tokenizer.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17.2M/17.2M [00:00<00:00, 24.8MB/s]
special_tokens_map.json: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 296/296 [00:00<00:00, 3.72MB/s]
chat_template.jinja: 4.61kB [00:00, 20.3MB/s]
[2025-10-16 15:13:53][src.model_setup][INFO] - ✓ Tokenizer loaded (vocab size: 128256)
[2025-10-16 15:13:53][src.model_setup][INFO] - ⚠️  Applying chat format setup (special tokens not found in tokenizer)...
[2025-10-16 15:13:53][src.model_setup][INFO] -   has_chat_template: True, vocab_already_extended: False
[2025-10-16 15:13:53][src.model_setup][INFO] -   has_chat_tokens: False
[2025-10-16 15:13:53][src.model_setup][WARNING] - This may cause embedding mismatch with the trained adapter!
/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/trl/models/utils.py:116: FutureWarning: The `setup_chat_format` function is deprecated and will be removed in version 0.26.0. Please use `clone_chat_template` instead.
  warnings.warn(
[2025-10-16 15:13:53][src.model_setup][ERROR] - ❌ Failed to load trained model: Chat template is already added to the tokenizer. If you want to overwrite it, please set it to None
[2025-10-16 15:13:53][src.model_setup][ERROR] - Traceback (most recent call last):
  File "/home/chrisjcc/text-to-sql-finetuning/src/model_setup.py", line 324, in load_trained_model
    model, tokenizer = setup_chat_format(model, tokenizer)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/trl/models/utils.py", line 123, in setup_chat_format
    raise ValueError(
ValueError: Chat template is already added to the tokenizer. If you want to overwrite it, please set it to None

Error executing job with overrides: []
Traceback (most recent call last):
  File "/home/chrisjcc/text-to-sql-finetuning/scripts/evaluate.py", line 630, in main
    evaluator.load_model(adapter_path=adapter_for_examples)
  File "/home/chrisjcc/text-to-sql-finetuning/scripts/evaluate.py", line 239, in load_model
    self.model, self.tokenizer = ModelSetup.load_trained_model(
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chrisjcc/text-to-sql-finetuning/src/model_setup.py", line 324, in load_trained_model
    model, tokenizer = setup_chat_format(model, tokenizer)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/chrisjcc/miniconda3/envs/llm_env/lib/python3.11/site-packages/trl/models/utils.py", line 123, in setup_chat_format
    raise ValueError(
ValueError: Chat template is already added to the tokenizer. If you want to overwrite it, please set it to None

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
⚠ Evaluation failed.
make: *** [Makefile:60: evaluate] Error 1
```

--- a/scripts/evaluate.py
+++ b/scripts/evaluate.py
@@ -196,7 +196,9 @@ class ModelEvaluator:
         test_dataset: Dataset,
         batch_size: int = 8,
         temperature: float = 0.0,
-        skip_baseline: bool = False
+        skip_baseline: bool = False,
+        setup_chat_format: bool = True,
+        force_chat_setup: bool = False,
     ):
         """
         Initialize evaluator.
@@ -208,6 +210,8 @@ class ModelEvaluator:
             batch_size: Batch size for generation
             temperature: Sampling temperature for generation (0.0 = greedy)
             skip_baseline: If True, skip baseline evaluation
+            setup_chat_format: Whether to apply chat format setup (default: True)
+            force_chat_setup: If True, force chat setup even if template exists (default: False)
         """
         self.base_model_path = model_name_or_path
         self.adapter_path = adapter_path
@@ -215,6 +219,8 @@ class ModelEvaluator:
         self.batch_size = batch_size
         self.temperature = temperature
         self.skip_baseline = skip_baseline
+        self.setup_chat_format = setup_chat_format
+        self.force_chat_setup = force_chat_setup
         self.model = None
         self.tokenizer = None
         self.device = "cuda" if torch.cuda.is_available() else "cpu"
@@ -238,7 +244,9 @@ class ModelEvaluator:
 
         self.model, self.tokenizer = ModelSetup.load_trained_model(
             model_path=self.base_model_path,
-            adapter_path=adapter_path
+            adapter_path=adapter_path,
+            setup_chat_format_flag=self.setup_chat_format,
+            force_chat_setup=self.force_chat_setup,
         )
 
         # Merge adapter into base model if it's a PEFT model
@@ -623,6 +631,8 @@ def main(cfg: DictConfig):
         batch_size=cfg.evaluation.batch_size,
         temperature=cfg.evaluation.temperature,
         skip_baseline=cfg.evaluation.skip_baseline,
+        setup_chat_format=cfg.hf.setup_chat_format,
+        force_chat_setup=cfg.hf.force_chat_setup,
     )
 
     # Show examples (load adapter if available, otherwise base model)

--- a/scripts/inference.py
+++ b/scripts/inference.py
@@ -316,7 +316,9 @@ def main(cfg: DictConfig):
     try:
         model, tokenizer = ModelSetup.load_trained_model(
             model_path=model_path,
-            adapter_path=adapter_path
+            adapter_path=adapter_path,
+            setup_chat_format_flag=cfg.hf.setup_chat_format,
+            force_chat_setup=cfg.hf.force_chat_setup,
         )
         if hasattr(model, "merge_and_unload"):
             model = model.merge_and_unload()

--- a/src/model_setup.py
+++ b/src/model_setup.py
@@ -172,6 +172,8 @@ class ModelSetup:
         model_path: str,
         adapter_path: Optional[str] = None,
         device_map: str = "auto",
+        setup_chat_format_flag: bool = True,
+        force_chat_setup: bool = False,
     ) -> Tuple[PreTrainedModel, PreTrainedTokenizer]:
         """
         Load a trained PEFT model (adapter + tokenizer) for inference.
@@ -180,6 +182,8 @@ class ModelSetup:
             model_path: Base model path or HuggingFace ID
             adapter_path: Optional path to adapter (local or HuggingFace Hub ID). If None, loads base model only.
             device_map: Device mapping strategy
+            setup_chat_format_flag: Whether to apply chat format setup (default: True)
+            force_chat_setup: If True, force chat setup even if template exists (default: False)
 
         Returns:
             Tuple of (model, tokenizer)
@@ -298,9 +302,19 @@ class ModelSetup:
             has_chat_tokens = any(token in tokenizer.get_vocab() for token in chat_special_tokens)
             vocab_already_extended = tokenizer_vocab_size > model.config.vocab_size
 
-            # Only skip setup_chat_format if we have BOTH chat template AND extended vocab
-            # This ensures we use the tokenizer exactly as it was saved during training
-            if has_chat_template and vocab_already_extended:
+            # Check if we should apply chat format setup based on configuration
+            if not setup_chat_format_flag:
+                # User explicitly disabled chat format setup via configuration
+                logger.info("⊗ Skipping chat format setup (setup_chat_format=false in config)")
+                logger.info(f"  Tokenizer vocab: {tokenizer_vocab_size}, Model vocab: {model.config.vocab_size}")
+                logger.info(f"  Has chat template: {has_chat_template}, Has chat tokens: {has_chat_tokens}")
+
+                # Resize model embeddings to match tokenizer if needed
+                if tokenizer_vocab_size != model.config.vocab_size:
+                    logger.info(f"Resizing model embeddings from {model.config.vocab_size} to {tokenizer_vocab_size}")
+                    model.resize_token_embeddings(tokenizer_vocab_size, mean_resizing=False)
+                    logger.info("✓ Model embeddings resized (preserving adapter's trained embeddings)")
+            elif has_chat_template and vocab_already_extended:
                 # Tokenizer already has chat format configured from training and vocab was extended
                 logger.info("✓ Chat format already configured in tokenizer (skipping setup_chat_format)")
                 logger.info(f"  Tokenizer vocab: {tokenizer_vocab_size}, Model vocab: {model.config.vocab_size}")
@@ -321,6 +335,12 @@ class ModelSetup:
                 logger.info(f"  has_chat_template: {has_chat_template}, vocab_already_extended: {vocab_already_extended}")
                 logger.info(f"  has_chat_tokens: {has_chat_tokens}")
                 logger.warning("This may cause embedding mismatch with the trained adapter!")
+
+                # Handle force_chat_setup flag
+                if force_chat_setup and has_chat_template:
+                    logger.warning("Force chat setup enabled - setting existing chat template to None")
+                    tokenizer.chat_template = None
+
                 model, tokenizer = setup_chat_format(model, tokenizer)
                 new_vocab_size = len(tokenizer)
                 logger.info(f"✓ Chat format applied (new vocab size: {new_vocab_size})")

